File: 2537-count-the-number-of-good-subarrays/2537-count-the-number-of-good-subarrays.py
from collections import defaultdict

class Solution:
    def countGood(self, nums: List[int], k: int) -> int:
        freq = defaultdict(int)
        left = 0
        pair_count = 0
        good_subarrays = 0

        for right in range(len(nums)):
            pair_count += freq[nums[right]]
            freq[nums[right]] += 1

            while pair_count >= k:
                good_subarrays += len(nums) - right
                freq[nums[left]] -= 1
                pair_count -= freq[nums[left]]
                left += 1

        return good_subarrays

File: 2537-count-the-number-of-good-subarrays/2537-count-the-number-of-good-subarrays.cpp
class Solution {
public:
    long long countGood(vector<int>& nums, int k) {
        int n = nums.size();
        int i=0, j = 0;
        unordered_map<int,int>mp;
        long long res =0,pairs=0;
        while(j<n){
            pairs+=mp[nums[j]];
            mp[nums[j]]++;
            while(pairs>=k){
                res+=n-j;
                mp[nums[i]]--;
                pairs-=mp[nums[i]];
                i++;
            }
            j++;
        }
        return res;
    }
};